File: coding-fibonacci.py
# ref: https://devdoggo.netlify.com/post/alg_ds/challenges/fibonacci_solutions/
import time

def call_counter(func):
  def helper(x):
    helper.calls += 1
    return func(x)
  helper.calls = 0
  return helper

@call_counter
def fibonacci_calc(num):
  if num < 2:
    return num

  n1 = 1
  n2 = 1
  result = 0
  for i in range(3, num+1):
    result = n1 + n2
    n1, n2 = n2, result
  return result

@call_counter
def fibonacci_recursive(num):
  if num < 2:
    return num
  return fibonacci_recursive(num-1) + fibonacci_recursive(num-2)

memo = {1:1, 2:1}
@call_counter
def fibonacci_recursive_memo(num):
  if num == 0:
    return 0
  if num not in memo:
    # print("not in", num)
    memo[num] = fibonacci_recursive_memo(num-1) + fibonacci_recursive_memo(num-2)
  # else:
  #   print("in", num)
  return memo[num]

@call_counter
def fibonacci_python(num):
  a, b = 1, 0
  for i in range(num):
    a, b = b, a + b
  return b

if __name__ == "__main__":
  num = int(input("Enter number for fibonacci> "))
  print()

  start_time = time.time()
  fibonacci_number = fibonacci_calc(num)
  end_time = time.time()
  print("function call count(calc):", fibonacci_calc.calls)
  print("elaspsed time(calc):", end_time - start_time)
  print("result:", fibonacci_number)
  print()

  start_time = time.time()
  fibonacci_number = fibonacci_recursive(num)
  end_time = time.time()
  print("function call count(recursive):", fibonacci_recursive.calls)
  print("elaspsed time(recursive):", end_time - start_time)
  print("result:", fibonacci_number)
  print()

  start_time = time.time()
  fibonacci_number = fibonacci_recursive_memo(num)
  end_time = time.time()
  print("function call count(memo):", fibonacci_recursive_memo.calls)
  print("elaspsed time(memo):", end_time - start_time)
  print("result:", fibonacci_number)
  # print(memo)
  print()

  start_time = time.time()
  fibonacci_number = fibonacci_python(num)
  end_time = time.time()
  print("function call count(python):", fibonacci_python.calls)
  print("elaspsed time(python):", end_time - start_time)
  print("result:", fibonacci_number)

File: coding-fibonacci.go
package main

import (
	"fmt"
	"time"
)

// ref: https://github.com/godsman-yang/coding-learn/blob/master/coding-fibonacci.py

func fibonacciCalc(num int) int {
	Counter++ // global varialble

	if num < 2 {
		return num
	}

	n0 := 0
	n1 := 1
	// n2 := 1
	result := 0

	// for i in range(3, num+1):
	for i := 2; i <= num; i++ {
		result = n0 + n1
		n0, n1 = n1, result
	}
	return result
}

func fibonacciRecursive(num int) int {
	Counter++

	if num < 2 {
		return num
	}
	result := fibonacciRecursive(num-1) + fibonacciRecursive(num-2)
	// return fibonacci_recursive(num-1) + fibonacci_recursive(num-2)
	return result
}

//global variable
var Memo map[int]int = map[int]int{0: 0, 1: 1}

func fibonacciRecursiveMemo(num int) int {
	Counter++

	if num < 2 {
		return Memo[num]
	}

	_, ok := Memo[num]
	if ok == false {
		Memo[num] = fibonacciRecursiveMemo(num-1) + fibonacciRecursiveMemo(num-2)
	}
	return Memo[num]
}

func fibonacciGolang(num int) int {
	Counter++

	a, b := 1, 0
	for i := 0; i < num; i++ {
		a, b = b, a+b
	}
	return b
}

// global variables
var Counter int

func main() {
	num := InputFibonacciNumber()
	fmt.Println("")

	Counter = 0
	startTime := time.Now()
	fibonacciNumber := fibonacciCalc(num)
	elapsedTime := time.Since(startTime)

	fmt.Println("Function call count(clac):", Counter)
	fmt.Printf("Elaspsed time(calc): %s\n", elapsedTime)
	fmt.Println("Result:", fibonacciNumber)
	fmt.Println()

	Counter = 0
	startTime = time.Now()
	fibonacciNumber = fibonacciRecursive(num)
	elapsedTime = time.Since(startTime)

	fmt.Println("Function call count(recursive):", Counter)
	fmt.Printf("Elaspsed time(recursive): %s\n", elapsedTime)
	fmt.Println("Result:", fibonacciNumber)
	fmt.Println()

	Counter = 0
	startTime = time.Now()
	fibonacciNumber = fibonacciRecursiveMemo(num)
	elapsedTime = time.Since(startTime)

	fmt.Println("Function call count(memo):", Counter)
	fmt.Printf("Elaspsed time(memo): %s\n", elapsedTime)
	fmt.Println("Result:", fibonacciNumber)
	fmt.Println()

	Counter = 0
	startTime = time.Now()
	fibonacciNumber = fibonacciGolang(num)
	elapsedTime = time.Since(startTime)

	fmt.Println("Function call count(golang):", Counter)
	fmt.Printf("Elaspsed time(golang): %s\n", elapsedTime)
	fmt.Println("Result:", fibonacciNumber)
	fmt.Println()
}

func InputFibonacciNumber() int {
	var number int

	fmt.Printf("Enter number for fibonacci> ")
	_, _ = fmt.Scanf("%d", &number)

	return number
}
